consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.consensus.ibft.statemachine;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import java.util.Set;
+import tech.pegasys.pantheon.consensus.ibft.ConsensusRoundIdentifier;
+import tech.pegasys.pantheon.consensus.ibft.IbftHelpers;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.RoundChangeCertificate;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.RoundChangePayload;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.SignedData;
+import tech.pegasys.pantheon.consensus.ibft.validation.RoundChangeMessageValidator;
+import tech.pegasys.pantheon.consensus.ibft.validation.RoundChangeMessageValidator.MessageValidatorFactory;
+import tech.pegasys.pantheon.ethereum.core.Address;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Responsible for handling all RoundChange messages received for a given block height
+ * (theoretically, RoundChange messages for a older height should have been previously discarded,
+ * and messages for a future round should have been buffered).
+ *
+ * <p>If enough RoundChange messages all targeting a given round are received (and this node is the
+ * proposer for said round) - a newRound message is sent, and a new round should be started by the
+ * controlling class.
+ */
+public class RoundChangeManager {
+
+  public static class RoundChangeStatus {
+    private final int quorumSize;
+
+    @VisibleForTesting
+    final Map<Address, SignedData<RoundChangePayload>> receivedMessages = Maps.newHashMap();
+
+    private boolean actioned = false;
+
+    public RoundChangeStatus(final int quorumSize) {
+      this.quorumSize = quorumSize;
+    }
+
+    public void addMessage(final SignedData<RoundChangePayload> msg) {
+      if (!actioned) {
+        receivedMessages.put(msg.getSender(), msg);
+      }
+    }
+
+    public boolean roundChangeReady() {
+      return receivedMessages.size() >= quorumSize && !actioned;
+    }
+
+    public RoundChangeCertificate createRoundChangeCertificate() {
+      if (roundChangeReady()) {
+        actioned = true;
+        return new RoundChangeCertificate(receivedMessages.values());
+      } else {
+        throw new IllegalStateException("Unable to create RoundChangeCertificate at this time.");
+      }
+    }
+  }
+
+  private static final Logger LOG = LogManager.getLogger();
+  @VisibleForTesting
```

Comment:
concerned about "VisibleForTesting" on internal data, both here and above.

Reply:
I'm open to suggestions of how to test without it.